### DISCUSSION TITLE
"Scaling up" physics and collision for larger resolutions/sprite sizes

### DISCUSSION THREAD

**Discussion by Candescence** [Techokami/SonicWorldsNext#14] [Q&A]
This is kind of a different thing than widescreen/zoom, because sometimes there'll be projects that'll want to have larger resolution sprite sizes - for example, Freedom Planet 2 went for 640x360 for its screen size (since it doubles to 720p and triples to 1080p). Like, many fangames will stick with the original resolution or widescreen like 424 x 240, but I personally want to work with something double or triple that size for my project.

So far I'm not having a ton of trouble adjusting certain physics values and changing hitbox sizes, though it's hard to figure out what exactly should be changed and what would do just fine staying the same. But also larger player characters is _glitchy,_ and I'm not sure why - the main glitches I've encountered are the player popping up on top of walls instead of being stopped by them, and crouching doesn't work, nevermind spin-dashing.

So if anyone has got some pointers on how to better go about scaling up the player without straight up breaking things it would be really appreciated.

**Comment by DimensionWarped** [comment]
I'm sure it's going to be a bit of a lift as things are since most things in the engine are built around the expectation of a specific resolution. I doubt it's ever going to be a goal of the project to make this an easy effort either.

What could be created though is a case study to determine the exact set of things that need to be changed to maintain basic functionality.

Perhaps you could create a fork (feel free to use dummy graphics if you don't want to show off the specifics of what you are working on) with assets at the target size you are looking at? I'd be willing to spend some time trying to figure it out if so.

**Reply by Candescence** [reply]
Honestly atm I'm just doubling the sprite size, doubling the size of the hitboxes and doubling the size of the tiles (along with adding Lilac's FP2 sprite as a size reference), pretty early days in terms of my own project right now so I haven't done much with it. But I'll see about making a fork, sure.

**Reply by DimensionWarped** [reply]
Did you turn on show collision in the debug menu? Might help you see if something really obvious is screwing up.

**Reply by Candescence** [reply]
I did, it doesn't seem like the collisions are at all different from the normal size... Except for when you try to crouch, instead being in midair rather than snapping to the ground, it's weird. Rolling works, though.

Also, speaking of rolling, rolling goes straight through walls, rather than immediately snapping to the top of them. Very peculiar.

**Reply by Candescence** [reply]
But yeah anyhow I've gone and added the new/modified stuff my fork: https://github.com/Candescence/SonicWorldsNext_DoubleSize

Just run the "DoubleZone", which is currently just a very, very simple zone with some slopes and a loop at the moment. Right now players can't complete the loop.

I also increased the acceleration, friction and jump height to account for the increased scale, probably should adjust the gravity as well since the jumping feels more floaty than it should.

**Comment by DimensionWarped** [comment]
I've spent about half an hour on this so far without figuring out what magic combination of things are wrong. Though I have made some corrections to hitboxes that were *obviously* wrong, none of them have had the impact I was hoping for.

What I can tell you though, is that the reason your game is loading into BazeZone instead of DoubleZone in spite of having written DoubleZone into a number of scripts at this point is because the next zone value is exported in the character select script, so you need to change it here:

![fix_it_here](https://user-images.githubusercontent.com/7477617/216865610-4baafbe4-75cb-415a-861e-2ec20b15d172.png)

If that was confusing you as much as it was confusing me, you're welcome. If I'm mistaken, then... well, sorry for assuming you were having a problem that you weren't.

I'll keep poking at this for a while.

**Reply by Candescence** [reply]
Honestly that was _absolutely_ confusing the absolute hell out of me, thank you _so much._ Probably also explains why the project was loading BaseZone even when it was deleted from the project!

**Reply by DimensionWarped** [reply]
I'd strongly recommend learning grep if you don't use it already. I don't think I would have been able to figure this out without it. But grepping for 'BaseZone' led me to it rather quickly.

**Comment by DimensionWarped** [comment]
*sigh*

I've been twiddling knobs for about an hour now and can't for the life of me figure out any of the major physics issues. I'll defer to @Renhoex + @Techokami on this one.

I'd love to get an explainer. I've been poking around mostly at PhysicsObject.gd and Player.gd as well as looking over their linked tscn files trying to find any obvious areas where something is dependent on magic numbers mostly.

**Reply by Candescence** [reply]
No worries, at least you gave it a shot, mate. I too would like to know the finer points of how to adjust the hitboxes to work properly at different character sizes without breaking things. Who knows, maybe this little flight of fancy of mine might help make Worlds Next more robust.

**Reply by Renhoex** [reply]
Hi I've been looking over the code and I was able to fix up a lot of the issues, I'll try to break it down into parts.

Firstly I wanna make a recommendation,
By default the project is set to viewport, viewport scaling is for pixel perfect graphics but I saw you mentioned that you wanted to implement zooming, if that's the case I recommend setting the scaling mode to 2D that way your sprites don't get crunched.
![scaling](https://user-images.githubusercontent.com/26295900/216899024-4dc63a67-c531-4731-8ffa-15db7dd04241.PNG)
I'd like to mention this as an alternative solution since doubling the size might lead to a few issues. But here's an example, i set lilacs scale to 0.5, here's what she looks like in game in viewport, vs 2D
![ViewportMode](https://user-images.githubusercontent.com/26295900/216899257-2c665517-f4ee-44cf-a562-6c4437640e2b.PNG)
![2DView](https://user-images.githubusercontent.com/26295900/216899272-df88749d-f00a-459d-95fa-48323466fb3a.PNG)


**Wall Clipping Fix**
This one's pretty easy, so PhysicsObject.gd pulls the players mask width at the start of the scene for the push radius, you've done right in setting the player hitboxes through the code but you didn't set this in the collisionshape2D shape so that one still uses the old one (I'll be sure to update this so that it uses the code's in future)
It's an easy fix just set the player's collision shape to match your codes, that'll stop them clipping into walls.
![hitbox](https://user-images.githubusercontent.com/26295900/216899794-b17510c2-ccf4-4fab-9c56-5dc37fca4258.PNG)


**Collision issues**
I noticed in your Big map that you didn't set the collision mask bits, currently it's only set as a "top solid", masks help determine how the player interacts with the level, this stuff is going to be one of the things we cover when we get around to getting the documentation going.
![Collissions](https://user-images.githubusercontent.com/26295900/216900079-a6901a55-7866-4a29-b94c-463ff1cf02c7.png)
![CollissionsB](https://user-images.githubusercontent.com/26295900/216900073-b7da83fa-3f50-427b-b955-e77ac9a4bcc0.PNG)

Be sure to keep us updated if any other issues come up, I'll be sure to keep this stuff in mind going into updates going forward so that other people don't encounter the same issue.

**Reply by Candescence** [reply]
@Renhoex You absolute legend, thank you! That definitely helps a ton.

Though there still is some weird hitbox issues, as seen here:

![image](https://user-images.githubusercontent.com/9062079/216904700-e89aa1b3-67d2-4123-94fc-13b337d35c83.png)
![image](https://user-images.githubusercontent.com/9062079/216906060-5acb1c98-8269-4354-9496-d900ab8093e8.png)

It's like the hitbox is stuck in "rolling mode", it reverts to normal when landing from a jump but then immediately changes. It also still causes crouching to be buggy, like it's flipping between two hitbox modes that it can't decide on. It's weird. Any particular hitbox settings that I'm doing wrong?

**Reply by Renhoex** [reply]
I'm not sure what's causing this? Everything I'm doing worked fine on my build.
did you change anything between when you made the fork and when you did these changes?
like did you change the hitboxes or the sprite's offsets?
for reference this is how it looks on mine
`const HITBOXESSONIC = {NORMAL = Vector2(18,38), ROLL = Vector2(14,28), CROUCH = Vector2(18,22), GLIDE = Vector2(20,20)}`

**Reply by Candescence** [reply]
> I'm not sure what's causing this? Everything I'm doing worked fine on my build. did you change anything between when you made the fork and when you did these changes? like did you change the hitboxes or the sprite's offsets? for reference this is how it looks on mine `const HITBOXESSONIC = {NORMAL = Vector2(18,38), ROLL = Vector2(14,28), CROUCH = Vector2(18,22), GLIDE = Vector2(20,20)}`

Okay, so the script is the same on my end, so I realized I might've accidentally messed something up when figuring out how to change the hitbox extents, so I restored the old Player 'prefab' and changed the extents again, without messing with the original rectangle bounding box otherwise, and that fixed it.

The crouch is still bugged, though, seems to be unable to properly finish the crouch animation before it somehow resets to standing.

Anyhow, pushed the fixes to the fork.

**Reply by Renhoex** [reply]
The crouching bug happens in my build so I'll take a look at that when I get time

**Reply by Renhoex** [reply]
Ok so this is part of the scaling stuff, it requires a few changes.

**Mask**
I had to increase the mask of the crouch because it would be too small and make the player clip into the floor,
`CROUCH = Vector2(18,30)`

**Physics Object**
you need to change the code for the physics object as the sensors are too small, open the script at `res://Scripts/Objects/PhysicsObject.gd`
at line 21, change 14 to 14*2 for groundLookDistance
at line 92, change the +4 to +8
`	var extendFloorLook = min(abs(movement.x/60)+8,groundLookDistance)*(int(movement.y >= 0)*int(ground))`

**Reply by Candescence** [reply]
> Ok so this is part of the scaling stuff, it requires a few changes.
> 
> **Mask** I had to increase the mask of the crouch because it would be too small and make the player clip into the floor, `CROUCH = Vector2(18,30)`
> 
> **Physics Object** you need to change the code for the physics object as the sensors are too small, open the script at `res://Scripts/Objects/PhysicsObject.gd` at line 21, change 14 to 14*2 for groundLookDistance at line 92, change the +4 to +8 ` var extendFloorLook = min(abs(movement.x/60)+8,groundLookDistance)*(int(movement.y >= 0)*int(ground))`

Okay, that fixes the crouching bug, thank you.

So the only problems left that I can tell right now is that Tails still has the standing/walking/running short hitbox issue, and Knuckles' wall-climbing doesn't work - he latches on to wall momentarily before letting go.

![image](https://user-images.githubusercontent.com/9062079/216987304-1ad112c0-fe6f-46c4-a359-e2038438f52f.png)

**Reply by Renhoex** [reply]
Yeah this is why scaling up might have some issues, unfortunately I'm probably not going to be able to fix this until tomorrow, I'll take a look and see what I can do as soon as I can.

It may be an offsetting issue in tail's hitbox, most I can really say is experiment with the sprite position or mask sizes for now

**Reply by DimensionWarped** [reply]
> **Collision issues** I noticed in your Big map that you didn't set the collision mask bits, currently it's only set as a "top solid", masks help determine how the player interacts with the level, this stuff is going to be one of the things we cover when we get around to getting the documentation going.

I'm embarrassed at not catching this one. I had actually suspected it since I was able to clear loops in the normal zone just fine, but I was having trouble figuring out where the collision layers were set on tiles maps (though it seems obvious now that I know it's what I should be looking for) and had a hard time imagining that this might not have been a straight copy and had no reason to lose its collision.

I'll start a wiki document. Probably won't be able to finish it, but I think I understand layers 1,2,3,4 + 5,6,7,8/9,10,11,12 + 14 and 15 at this point.

**Comment by DimensionWarped** [comment]
Sonic values:
const HITBOXESSONIC = {NORMAL = Vector2(18,39), ROLL = Vector2(14,28), CROUCH = Vector2(18,32), GLIDE = Vector2(20,20)}

Tails values:
const HITBOXESTAILS = {NORMAL = Vector2(18,35), ROLL = Vector2(14,26), CROUCH = Vector2(18,30), GLIDE = Vector2(20,20)}

Give those a shot and tell me how you like them. I'll do Knuckles a little later.

**Reply by Candescence** [reply]
That fixes the Tails hitbox issue, at least. Though there is a _slight_ shift in the camera when crouching - not a big deal, but it's noticeable. Setting Sonic's hitbox height when crouching to 31 and Tails' to 29 fixes it.

**Reply by DimensionWarped** [reply]
Strange, I'm not getting that. We'll have to compare branches at some point soon.

EDIT: Well, once I make one anyway. Currently all my stuff is just local.

**Reply by Candescence** [reply]
I pushed all my changes to the main branch of the fork, so you can take a look yourself.

**Reply by Renhoex** [reply]
took me a while because it's been a while since I looked at the code, I couldn't replicate the issue you were talking about when crouching but I did notice the camera shifts when you start rolling.

if you wanna fix that you'll want to go into the player script and look at line 1128 and change that chunk of code to this

#
	match(character):
		CHARACTERS.TAILS:
			match($HitBox.shape.extents):
				currentHitbox.ROLL:
					camAdjust = Vector2(0,-3)
				_:
					camAdjust = Vector2.ZERO
		_: # default
			match($HitBox.shape.extents):
				currentHitbox.ROLL:
					camAdjust = Vector2(0,-11)
				_:
					camAdjust = Vector2.ZERO

**Reply by Candescence** [reply]
> took me a while because it's been a while since I looked at the code, I couldn't replicate the issue you were talking about when crouching but I did notice the camera shifts when you start rolling.
> 
> if you wanna fix that you'll want to go into the player script and look at line 1128 and change that chunk of code to this
> 
> ```
> match(character):
> 	CHARACTERS.TAILS:
> 		match($HitBox.shape.extents):
> 			currentHitbox.ROLL:
> 				camAdjust = Vector2(0,-3)
> 			_:
> 				camAdjust = Vector2.ZERO
> 	_: # default
> 		match($HitBox.shape.extents):
> 			currentHitbox.ROLL:
> 				camAdjust = Vector2(0,-11)
> 			_:
> 				camAdjust = Vector2.ZERO
> ```

I actually had to set Tails' roll adjustment to 9 to fix the problem, in this case, weirdly enough.

**Comment by DimensionWarped** [comment]
BTW, I'm really liking the 640x360 resolution as a conceptual resolution. You're right about it being ideal for scaling to everything normal, the only issue with it for a Sonic game (using Sonic 3-derived assets anyway) is that it's a little *too* zoomed out. Reminds me of playing Jazz Jackrabbit 2 and the level of zoom of that game just took me right out of the action. That said, I think doubling the size is going to lead to it feeling screen crunched. I don't know how plausible it is for us to get sprites that scale everything up by about 50%, but that feels like it would be the golden zone for this resolution to me.

**Comment by Candescence** [comment]
So now that Next 2.0 is out and I've been dabbing my toes back into gamedev, I've gotten back into recreating the 2x tileset with the new Godot 4 tileset system...

And I just realized that I have to actually create all the collisions. By hand. For _all four collision layers._ Oh God, that's gonna be a miserable experience. I might actually do what Renhoex suggested and stick to just using half-sized sprites in 2D/canvas_items mode for my project for the time being.

**Comment by Renhoex** [comment]
yeah godot 4's new tile system still needs a lot of work, my current suggestion is to build up chunks in the tile map or use the base tileset and turn off visibility when playing the stage.

**Reply by Candescence** [reply]
Kinda, I suppose? I'm thinking of maybe just duplicating the base tileset when I'm done with the tile collisions, and just have multiple tilemaps for layers, which would be, as you say, invisible during gameplay. It's not quite as good, but it'd save me a lot of headaches having to duplicate my work.

**Reply by DimensionWarped** [reply]
![image](https://github.com/Techokami/SonicWorldsNext/assets/7477617/7e5bc980-da94-4192-8166-b3544ea888d9)

**Reply by Candescence** [reply]
I'm pretty sure that's only doable with a single tileset, though, and you can't assign different tile layers with specific collision layers from what I can tell, it _has_ to be done in the tileset

**Reply by Renhoex** [reply]
ah yeah, well that's why I set up the base tiles with the red and blue ones being low and high tile collisions

**Reply by Candescence** [reply]
GOOD NEWS I FIGURED OUT A SOLUTION.

So, I recalled that a lot of Godot's asset formats are just plaintext files, so I took a look in Notepad++. This is what the very first tile looks like, data-wise:

1:0/0 = 0
1:0/0/physics_layer_0/linear_velocity = Vector2(0, 0)
1:0/0/physics_layer_0/angular_velocity = 0.0
1:0/0/physics_layer_0/polygon_0/points = PackedVector2Array(-16, -16, 16, -16, 16, 16, -16, 16)
1:0/0/physics_layer_1/linear_velocity = Vector2(0, 0)
1:0/0/physics_layer_1/angular_velocity = 0.0
1:0/0/physics_layer_2/linear_velocity = Vector2(0, 0)
1:0/0/physics_layer_2/angular_velocity = 0.0
1:0/0/physics_layer_3/linear_velocity = Vector2(0, 0)
1:0/0/physics_layer_3/angular_velocity = 0.0

After backing up the tilemap, of course, I decided to experiment - I just changed "1:0/0/physics_layer_0/polygon_0/points = PackedVector2Array(-16, -16, 16, -16, 16, 16, -16, 16)" in the high tilemap to "1:0/0/physics_layer_1/polygon_0/points = PackedVector2Array(-16, -16, 16, -16, 16, 16, -16, 16)".

It worked. I just mass-replaced the physics layer number for each tile's polygon. Just took a couple of minutes to do. It all works exactly as one would expect.

So now I have all four layers of tiles fully set up. Setting up the first one was actually easier than I initially realized, since the tileset editor lets you edit multiple tiles at once, including tile collisions, and considering the number of similar but flipped/rotated tiles there were, it was pretty easy to just create polygons for entire sets of tiles at once and just individually flip/rotate them. So it turned out only the foreground layer needed the painstaking work to set up.

So yeah, nevermind, I'm good now with the tiles. Whew.

**Reply by Renhoex** [reply]
Glad to hear!
